lib/auth/agent_version_report.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/autoupdate"
+	"github.com/gravitational/teleport/lib/inventory"
+)
+
+type instanceReport struct {
+	data      map[string]instanceGroupReport
+	timestamp time.Time
+}
+
+func (ir instanceReport) collectInstance(handle inventory.UpstreamHandle) {
+	// If the instance is being soft-reloaded or shut down, we ignore it.
+	if goodbye := handle.Goodbye(); goodbye != nil && (goodbye.GetSoftReload() || goodbye.GetDeleteResources()) {
+		return
+	}
+
+	// We skip servers that joined less than a minute ago as they might have been
+	// connected to another auth instance a few seconds ago, which would lead to double-counting.
+	if handle.RegistrationTime().After(ir.timestamp.Add(-time.Minute)) {
+		return
+	}
+	// We skip auth and proxy instances because we don't update them.
+	if handle.HasService(types.RoleAuth) || handle.HasService(types.RoleProxy) {
+		return
+	}
+
+	hello := handle.Hello()
+
+	// If the machine has no updater, we skip it
+	if hello.ExternalUpgrader == "" {
+		return
+	}
+
+	// Reject instances who are not advertising the group properly.
+	// They might be running too old versions.
+	updateGroup := hello.GetUpdaterInfo().UpdateGroup
+	if updateGroup == "" {
+		return
+	}
+
+	// We skip instances whose updater status is not unknown or OK.
+	// Note: is it OK to allow unknown? Discuss this with Stephen.
+	status := hello.GetUpdaterInfo().UpdaterStatus
+	if status != types.UpdaterStatus_UPDATER_STATUS_OK && status != types.UpdaterStatus_UPDATER_STATUS_UNSPECIFIED {
+		return
+	}
+
+	if _, ok := ir.data[updateGroup]; !ok {
+		ir.data[updateGroup] = instanceGroupReport{}
+	}
+
+	ir.data[updateGroup].collectInstance(handle)
+}
+
+type instanceGroupReport map[string]instanceGroupVersionReport
+
+func (ir instanceGroupReport) collectInstance(handle inventory.UpstreamHandle) {
+	hello := handle.Hello()
+
+	// Note: not validating if the semver is correct, is this an issue?
+	stats, ok := ir[hello.Version]
+	if !ok {
+		stats = instanceGroupVersionReport{}
+	}
+
+	stats.count += 1
+
+	ir[hello.Version] = stats
+}
+
+type instanceGroupVersionReport struct {
+	count int
+	// Leaving room here to add the lowest UUID, as described in RFD 184.
+}
+
+func (a *Server) generateAgentVersionReport(ctx context.Context) (*autoupdatev1pb.AutoUpdateAgentReport, error) {
+	now := a.clock.Now()
+
+	a.logger.DebugContext(ctx, "Collecting agent versions from inventory")
+	rawreport := instanceReport{timestamp: now, data: make(map[string]instanceGroupReport)}
+	a.inventory.AllHandles(rawreport.collectInstance)
+
+	a.logger.DebugContext(ctx, "Building the agent version report")
+	spec := &autoupdatev1pb.AutoUpdateAgentReportSpec{
+		Timestamp: timestamppb.New(a.clock.Now()),
+		Groups:    make(map[string]*autoupdatev1pb.AutoUpdateAgentReportSpecGroup, len(rawreport.data)),
+	}
+
+	// TODO: cap the group and version map size
```

Comment:
How many groups do we expect? Just a handful, right?

Reply:
Yeah, cloud supports 5 groups by default./ On self hosted I would not expect more than 50 groups. Because this group info is configured by users and reported by user, we don't know exactly gome many groups are going to be reported. To avoid the resource growing too large I capped at 50 groups and 20 version per group.